mirrord/cli/src/port_forward.rs

```diff
@@ -0,0 +1,689 @@
+            listeners,
+            oneshots: VecDeque::new(),
+            sockets: HashMap::new(),
+            tasks: HashMap::new(),
+            internal_msg_tx,
+            internal_msg_rx,
+            waiting_for_pong: false,
+            ping_pong_timeout: Instant::now(),
+        })
+    }
+
+    pub(crate) async fn run(&mut self) -> Result<(), PortForwardError> {
+        // setup agent connection
+        let Ok(_) = self
+            .agent_connection
+            .sender
+            .send(ClientMessage::SwitchProtocolVersion(
+                mirrord_protocol::VERSION.clone(),
+            ))
+            .await
+        else {
+            // failed to send to agent
+            return Err(PortForwardError::AgentSetupError(
+                "Sending 'Switch Protocol Version' Client Message failed".into(),
+            ));
+        };
+        match self.agent_connection.receiver.recv().await {
+            Some(DaemonMessage::SwitchProtocolVersionResponse(version))
+                if CLIENT_READY_FOR_LOGS.matches(&version) =>
+            {
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::ReadyForLogs)
+                    .await
+                else {
+                    return Err(PortForwardError::AgentSetupError(
+                        "Sending 'Ready for Logs' Client Message failed".into(),
+                    ));
+                };
+            }
+            _ => {
+                return Err(PortForwardError::AgentSetupError(
+                    "Switching protocol version check failed".into(),
+                ))
+            }
+        }
+        tracing::trace!("Port forwarding setup complete");
+
+        loop {
+            select! {
+                _ = tokio::time::sleep_until(self.ping_pong_timeout.into()) => {
+                    if self.waiting_for_pong {
+                        // no pong received before timeout
+                        break Err(PortForwardError::AgentError("Agent failed to respond to Ping".into()));
+                    }
+                    let _ = self.agent_connection.sender.send(ClientMessage::Ping).await;
+                    self.waiting_for_pong = true;
+                    self.ping_pong_timeout = Instant::now() + Duration::from_secs(30);
+                },
+
+                message = self.agent_connection.receiver.recv() => match message {
+                    Some(message) => self.handle_msg_from_agent(message).await?,
+                    None => {
+                        tracing::trace!("None message received, connection ended with agent");
+                        break Ok(())
+                    },
+                },
+
+                // stream coming from the user app
+                message = self.listeners.next() => match message {
+                    Some(message) => self.handle_listener_stream(message)?,
+                    None => break Ok(()),
```

Comment:
```suggestion
                    None => unreachable!("created listener sockets are never closed"),
```
Given that we check `parsed_mappings` non-emptiness in `PortForwarder::new`